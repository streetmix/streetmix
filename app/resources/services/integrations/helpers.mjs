import models from '../../../db/models/index.js'

const { User, UserConnections, sequelize } = models

/**
 finds the database record for the given user
 */
// might be better to do something besides return null
// we can actually probably get rid of this function if we want, but it might help
// keep some of the other code clean
export async function findUser (userId) {
  const user = await User.findOne({ where: { auth0Id: userId } })
  if (user === null) {
    return null
  }
  return user.dataValues
}

// if the provider we're trying to find exists
// return its index value or -1 if not found
export function addOrUpdateByProviderName (array, item) {
  const indexValue = array.findIndex(
    (_item) => _item.provider === item.provider
  )
  if (indexValue === -1) {
    array.push(item)
  } else {
    array[indexValue] = item
  }
}

/**
 given a user, check their identity data,
 depending on what data you find, add subscriber role

 TODO: make this less permissive(maybe add a status field or have logic depending on the provider), tested,
 and remove the role if we're confident we should (probably shouldn't worry about this for MVP)
 */
export async function syncAccountStatus (userId) {
  const databaseUser = await User.findOne({ where: { auth0Id: userId } })
  if (databaseUser.identities.some((identity) => 'user_id' in identity)) {
    databaseUser.addRole('SUBSCRIBER_1')
  }
}

/**
 * When a user has successfully connected their Streetmix account with a third-
 * party identity provider, record that connected identity with the user's
 * Streetmix data in the UserConnection table.
 *
 * @todo Remove first attempt at storing identity information as a JSON blob
 *        on a User model
 * @param {Object} account - User's Streetmix account information
 * @param {Object} profile - User's connected identity information
 * @returns {Promise}
 */
export function addUserConnection (account, profile) {
  const identities = account.identities
  const identity = {
    provider: profile.provider,
    user_id: profile.id,
    access_token: profile.access_token || undefined, // Coil
    refresh_token: profile.refresh_token || undefined // Coil
  }

  // if provider exists, should update that item
  // if not, should add to the object list
  // TODO: Deprecated
  addOrUpdateByProviderName(identities, identity)

  return sequelize.transaction(async (t) => {
    // This is a previous attempt, where we saved connected identity
    // information to a JSON field in the User model. A newer method (using
    // the UserConnection table) is used below. Both are currently used to
    // test concurrent data validity. In time, the User.update() method should
    // be removed.
    await User.update(
      {
        identities
      },
      { where: { auth0Id: account.auth0Id }, returning: true, transaction: t }
    )

    await UserConnections.findOrCreate({
      where: {
        user_id: account.id,
        provider: profile.provider
      },
      transaction: t
    })
    return await UserConnections.update(
      {
        provider_user_id: profile.id,
        metadata: profile
      },
      {
        where: {
          user_id: account.id,
          provider: profile.provider
        },
        returning: true,
        transaction: t
      }
    )
  })
}
