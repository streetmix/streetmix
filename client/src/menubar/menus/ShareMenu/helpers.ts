import type { UserState } from '~/src/types'
import type { StreetState } from '@streetmix/types'
import type { IntlShape } from 'react-intl'

export interface SocialShareProps {
  shareText: string
  shareUrl: string
}

/**
 * Returns the current page URL.
 * This may eventually change to a short URL generator - strt.mx
 */
export function getSharingUrl (): string {
  return window.location.href
}

export function getSharingMessage (
  street: StreetState,
  user: UserState,
  intl: IntlShape
): string {
  const signedIn = user.signedIn ?? false
  const userId = user.signInData?.userId ?? ''
  let message = ''

  if (typeof street.creatorId === 'string') {
    if (signedIn && street.creatorId === userId) {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.my-street',
            defaultMessage: 'Check out my street, {streetName}, on Streetmix!'
          },
          { streetName: street.name }
        )
      } else {
        message = intl.formatMessage({
          id: 'menu.share.messages.my-street-unnamed',
          defaultMessage: 'Check out my street on Streetmix!'
        })
      }
    } else {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street',
            defaultMessage:
              'Check out {streetName} by {streetCreator} on Streetmix!'
          },
          { streetName: street.name, streetCreator: `@${street.creatorId}` }
        )
      } else {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street-unnamed',
            defaultMessage:
              'Check out this street by {streetCreator} on Streetmix!'
          },
          { streetCreator: `@${street.creatorId}` }
        )
      }
    }
  } else {
    if (typeof street.name === 'string') {
      message = intl.formatMessage(
        {
          id: 'menu.share.messages.anonymous-creator-street',
          defaultMessage: 'Check out {streetName} on Streetmix!'
        },
        { streetName: street.name }
      )
    } else {
      message = intl.formatMessage({
        id: 'menu.share.messages.anonymous-creator-street-unnamed',
        defaultMessage: 'Check out this street on Streetmix!'
      })
    }
  }

  return message
}
