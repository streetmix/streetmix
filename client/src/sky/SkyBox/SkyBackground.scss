.sky-background {
  // Backgrounds don't need to have parallax scrolling, so they don't
  // need to be as wide as the clouds, which do scroll. But we can't
  // have it wider than the viewport, because there's a rendering bug
  // in Chrome where if stars.svg is in a div that's too wide, it's
  // rendered pixellated for some reason.
  position: absolute;
  left: 0;
  width: 100vw;
  height: 100%;
}

/* Give the react-spring animations smooth transitions */
.sky-background-color {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  will-change: opacity;
}
