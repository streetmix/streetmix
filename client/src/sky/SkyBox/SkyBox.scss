@import "../../../styles/variables";

.street-section-sky {
  position: absolute;
  left: 0;
  right: 0;
  top: -70px; // Space for scrolling down when gallery open
  height: calc(100% - 100px);
  pointer-events: none;
  opacity: 1;

  // In read-only mode, there's no UI, so sky needs some more room
  body.read-only & {
    height: calc(100% - 20px);
  }
}

.rear-clouds,
.front-clouds {
  position: absolute;
  left: 0;
  width: 100%;
  background-repeat: repeat-x;
  background-position: 0 0;
  transition: opacity $skybox-transition;
}

.rear-clouds {
  height: 250px;
  bottom: 205px;
  background-image: url("./sky-rear.svg");
  background-size: 250px 250px;
}

.front-clouds {
  height: 280px;
  bottom: 55px;
  background-image: url("./sky-front.svg");
  background-size: 250px 280px;
}

// A setting activates this class and turns on environmental animations
// This is so it can be disabled based on preference or performance reasons
.sky-animations {
  .rear-clouds {
    animation-name: rear-clouds-move;
    animation-duration: 60s;
    animation-timing-function: linear;
    animation-iteration-count: infinite;
  }

  .front-clouds {
    animation-name: front-clouds-move;
    animation-duration: 30s;
    animation-timing-function: linear;
    animation-iteration-count: infinite;
  }
}

@keyframes rear-clouds-move {
  from {
    background-position: 0 0;
  }

  to {
    background-position: 250px 0;
  }
}

@keyframes front-clouds-move {
  from {
    background-position: 0 0;
  }

  to {
    background-position: 250px 0;
  }
}

.sky-foreground {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  transition: opacity $skybox-transition;
  z-index: $z-01-foreground;
  pointer-events: none;
  opacity: 0;
}
