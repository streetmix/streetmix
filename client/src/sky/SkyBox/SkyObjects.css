.sky-background-objects {
  position: absolute;
  left: 0;
  width: 100vw;
  height: 100%;
}

.sky-background-objects > div {
  opacity: 0;
  transform: translateY(-14px);
  transition:
    opacity var(--skybox-transition),
    transform var(--skybox-transition);
}

.sky-background-object-enter-done {
  opacity: 1 !important;
  transform: translateY(0) !important;
}
