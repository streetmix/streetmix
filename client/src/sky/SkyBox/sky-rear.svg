<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 500 500" width="500" height="500">
  <linearGradient id="a" gradientUnits="userSpaceOnUse" x1="250" y1="237.091" x2="250" y2="67.177" gradientTransform="matrix(1 0 0 -1 0 240)">
    <stop offset=".08" stop-color="#7d9baa" stop-opacity="0" />
    <stop offset="1" stop-color="#7d9baa" stop-opacity=".22" />
  </linearGradient>
  <path d="M0 0v104.6c9.2-1.9 18.7-2.9 28.5-2.9 54.1 0 101 30.3 124.9 74.9 19.4-10 41.3-15.6 64.6-15.6 19.3 0 37.7 3.9 54.5 10.9 16.7-23.5 44.1-38.9 75.2-38.9 24.8 0 47.4 9.8 63.9 25.8 21-28.3 52.3-47.7 88.4-54.2V0H0z"
    fill="url(#a)" />
  <linearGradient id="b" gradientUnits="userSpaceOnUse" x1="250" y1="3.955" x2="250" y2="135.58" gradientTransform="matrix(1 0 0 -1 0 240)">
    <stop offset="0" stop-color="#b3d2df" stop-opacity=".38" />
    <stop offset="0.8" stop-color="#d2edfa" stop-opacity=".41" />
  </linearGradient>
  <path d="M411.6 158.8c-16.6-16-39.1-25.8-63.9-25.8-31 0-58.5 15.4-75.2 38.9-16.8-7-35.2-10.9-54.5-10.9-23.3 0-45.3 5.6-64.6 15.6-23.8-44.6-70.8-74.9-124.9-74.9-9.8 0-19.3 1-28.5 2.9V240h500V104.6c-36.1 6.5-67.4 26-88.4 54.2z"
    fill="url(#b)" />
  <linearGradient id="c" gradientUnits="userSpaceOnUse" x1="250" y1="260" x2="250" y2="0" gradientTransform="matrix(1 0 0 -1 0 500)">
    <stop offset="0" stop-color="#b3d2df" stop-opacity=".38" />
    <stop offset="0.75" stop-color="#cce0e7" stop-opacity="0.1" />
    <stop offset="1" stop-color="#d0d0d0" stop-opacity="0" />
  </linearGradient>
  <rect x="0" y="240" width="500" height="260" style="fill:url(#c);fill-rule:nonzero;" />
</svg>
