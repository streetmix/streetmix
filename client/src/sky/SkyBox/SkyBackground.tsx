import React from 'react'
import { useTransition, animated } from '@react-spring/web'
import { getSkyboxDef } from '..'
import './SkyBackground.scss'

interface SkyBackgroundProps {
  skybox: string
}

function SkyBackground (props: SkyBackgroundProps): React.ReactElement {
  const { skybox } = props

  const transitions = useTransition(skybox, {
    from: { opacity: 0 },
    enter: { opacity: 1 },
    leave: { opacity: 0 },
    config: {
      duration: 500
    }
  })

  return (
    <div className="sky-background">
      {transitions((style, item) => (
        <animated.div
          className="sky-background-color"
          style={{ ...style, ...getSkyboxDef(item).style }}
        />
      ))}
    </div>
  )
}

export default SkyBackground
