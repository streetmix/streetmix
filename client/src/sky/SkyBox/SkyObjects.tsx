import React from 'react'
import { TransitionGroup, CSSTransition } from 'react-transition-group'
import { images } from '../../app/load_resources'
import type { SkyboxDefinition } from '..'
import './SkyObjects.scss'

interface SkyObjectsProps {
  objects: SkyboxDefinition['backgroundObjects']
}

function SkyObjects ({ objects = [] }: SkyObjectsProps): React.ReactElement {
  return (
    <TransitionGroup className="sky-background-objects">
      {objects.map((object) => {
        const img = images.get(object.image)

        // Render only if asset is found
        if (img !== undefined) {
          return (
            <CSSTransition
              key={object.image}
              timeout={{
                enter: 0,
                exit: 500
              }}
              classNames="sky-background-object"
            >
              <div
                style={{
                  width: object.width,
                  height: object.height,
                  position: 'absolute',
                  left: `calc(${object.left * 100}% - (${object.width}px / 2)`,
                  top: `calc(${object.top * 100}% - (${object.height}px / 2)`
                }}
              >
                <img src={img.src} style={{ width: '100%', height: '100%' }} />
              </div>
            </CSSTransition>
          )
        }

        // Render nothing if asset is missing
        // We cannot pass `null` to <TransitionGroup>, it will throw an error
        // So we pass a Fragment, which must also have a key because it is
        // inside an array
        // TODO: Render placeholder ("dev") asset? That way it will be more
        // obvious that something is missing, and work on placement/size can
        // be done even if an actual asset isn't ready yet.
        return <React.Fragment key={object.image} />
      })}
    </TransitionGroup>
  )
}

export default SkyObjects
