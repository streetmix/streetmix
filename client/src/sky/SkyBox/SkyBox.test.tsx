import React from 'react'
import { vi } from 'vitest'

import { render } from '~/test/helpers/render'
import SkyBox from './SkyBox'

vi.mock('../skybox-defs.json', () => ({
  default: require('../__mocks__/skybox-defs.json')
}))

// Mock the `images` object.
// Note: in real life, this is a Map where the .get()
// method looks up an object value by the `id` key.
// The `src` property is normally a data-url.
vi.mock('../../app/load_resources', () => ({
  images: {
    get: (id: string) => ({
      src: 'bar.svg'
    })
  }
}))

describe('SkyBox', () => {
  it('renders', () => {
    const { asFragment } = render(<SkyBox scrollPos={0} />, {
      initialState: { street: { skybox: 'foo' } }
    })
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders with objects', () => {
    const { asFragment } = render(<SkyBox scrollPos={0} />, {
      initialState: { street: { skybox: 'bar' } }
    })
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders background animations', () => {
    const { container } = render(<SkyBox scrollPos={0} />, {
      initialState: {
        street: { skybox: 'bar' },
        flags: {
          SKY_ANIMATED_CLOUDS: { value: true }
        }
      }
    })
    expect(
      container.querySelector('section')?.className.includes('sky-animations')
    ).toEqual(true)
  })
})
