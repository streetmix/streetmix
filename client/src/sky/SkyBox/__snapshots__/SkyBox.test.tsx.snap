// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SkyBox > renders 1`] = `
<DocumentFragment>
  <section
    class="street-section-sky"
  >
    <div
      class="sky-background"
    >
      <div
        class="sky-background-color"
        style="opacity: 0;"
      />
    </div>
    <div
      class="rear-clouds"
      style="opacity: 0.85;"
    />
    <div
      class="front-clouds"
      style="opacity: 0.85;"
    />
    <div
      class="sky-foreground"
      style="opacity: 0;"
    />
  </section>
</DocumentFragment>
`;

exports[`SkyBox > renders with objects 1`] = `
<DocumentFragment>
  <section
    class="street-section-sky"
  >
    <div
      class="sky-background"
    >
      <div
        class="sky-background-color"
        style="opacity: 0;"
      />
    </div>
    <div
      style="width: 100px; height: 100px; position: absolute; left: calc(50% - (100px / 2); top: calc(50% - (100px / 2);"
    >
      <img
        src="bar.svg"
        style="width: 100%; height: 100%;"
      />
    </div>
    <div
      class="rear-clouds"
      style="opacity: 0.85;"
    />
    <div
      class="front-clouds"
      style="opacity: 0.85;"
    />
    <div
      class="sky-foreground"
      style="opacity: 0;"
    />
  </section>
</DocumentFragment>
`;
